File: python/infinicore/ops/deepseek_v4_fused_rope.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def deepseek_v4_fused_rope_(
    query: Tensor,
    key: Tensor | None,
    freqs_cis: Tensor,
    positions: Tensor,
    inverse: bool = False,
) -> tuple[Tensor, Tensor | None]:
    _infinicore.deepseek_v4_fused_rope_(
        query._underlying,
        None if key is None else key._underlying,
        freqs_cis._underlying,
        positions._underlying,
        inverse,
    )
    return query, key


def deepseek_v4_fused_rope(
    query: Tensor,
    key: Tensor | None,
    freqs_cis: Tensor,
    positions: Tensor,
    inverse: bool = False,
) -> tuple[Tensor, Tensor | None]:
    return deepseek_v4_fused_rope_(query, key, freqs_cis, positions, inverse)

File: src/infinicore/deepseek_v4_ops/deepseek_v4_fused_rope/deepseek_v4_fused_rope.cc
#include "infinicore/ops/deepseek_v4_fused_rope.hpp"

namespace infinicore::op {

void deepseek_v4_fused_rope_(Tensor query,
                             std::optional<Tensor> key,
                             const Tensor &freqs_cis,
                             const Tensor &positions,
                             bool inverse) {
    deepseek_v4_fused_rope_kernel_(query, key, freqs_cis, positions, inverse);
}

} // namespace infinicore::op
